### PR TITLE
Add V-JEPA for video classification model

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Adds a classification model compatible with the released pretrained classifiers on Something-Something-v2 and Diving48 datasets.

Epic Kitchen is a bit special, with 3 different classifiers on top of pooler (noun, verb, action), so it might require an additional model class to handle the case (I'm not sure if we have to add it, let's see on the V-JEPA usage).

cc @yonigozlan @koustuvsinha 

P.S. commit history is a bit weird cause I started working on it before model merge, but diff is OK

### CODE REVIEW

**Comment by qubvel** [comment]
run-slow: vjepa2

**Comment by qubvel** [comment]
run-slow: vjepa2

**Comment by qubvel** [comment]
run-slow: vjepa2

**Comment by qubvel** [comment]
run-slow: vjepa2

File: src/transformers/models/vjepa2/modeling_vjepa2.py [hunk 1]
```diff
@@ -536,17 +537,21 @@ def forward(
...
-        mask_keep = m.unsqueeze(-1).repeat(1, 1, x.size(-1))
-        all_x += [torch.gather(x, dim=1, index=mask_keep)]
+    all_masked_tensors = []
+    for mask in masks:
+        mask = mask.to(tensor.device)
+        mask_keep = mask.unsqueeze(-1).repeat(1, 1, tensor.size(-1))
+        all_masked_tensors += [torch.gather(tensor, dim=1, index=mask_keep)]
 
-    return torch.cat(all_x, dim=0)
+    return torch.cat(all_masked_tensors, dim=0)
```
Comment (qubvel) [inline src/transformers/models/vjepa2/modeling_vjepa2.py:554]: Fix for multi-gpu support (cc @ydshieh)

File: src/transformers/models/vjepa2/modeling_vjepa2.py [hunk 2]
```diff
@@ -735,49 +740,304 @@ def forward(
...
     base_model_prefix = "vjepa2"
     main_input_name = "pixel_values_videos"
     supports_gradient_checkpointing = True
-    _no_split_modules = ["VJEPA2Layer"]
+    _no_split_modules = [
+        "VJEPA2Layer",
+        "VJEPA2PoolerSelfAttentionLayer",
+        "VJEPA2PoolerCrossAttentionLayer",
+        "VJEPA2PredictorEmbeddings",
+    ]
```
Comment (qubvel) [inline src/transformers/models/vjepa2/modeling_vjepa2.py:1009]: to fix multi-gpu test as well

File: src/transformers/models/vjepa2/modeling_vjepa2.py [hunk 3]
```diff
@@ -900,4 +1160,93 @@ def get_vision_features(self, pixel_values_videos) -> torch.Tensor:
...
+
+        ```python
+        >>> import torch
+        >>> import numpy as np
+        >>> from transformers import AutoVideoProcessor, VJEPA2ForVideoClassification
+
+        >>> device = "cuda"
+
+        >>> video_processor = AutoVideoProcessor.from_pretrained("facebook/vjepa2-vitl-fpc64-256")
+        >>> model = VJEPA2ForVideoClassification.from_pretrained("facebook/vjepa2-vitl-fpc64-256").to(device)
```
Comment (qubvel) [inline src/transformers/models/vjepa2/modeling_vjepa2.py]:
```suggestion
        >>> video_processor = AutoVideoProcessor.from_pretrained("facebook/vjepa2-vitl-fpc16-256-ssv2")
        >>> model = VJEPA2ForVideoClassification.from_pretrained("facebook/vjepa2-vitl-fpc16-256-ssv2").to(device)
```

File: tests/models/vjepa2/test_modeling_vjepa2.py [hunk 4]
```diff
@@ -267,7 +267,7 @@ def test_inference_image(self):
             [[-0.0061, -1.8365, 2.7343], [-2.5938, -2.7181, -0.1663], [-1.7993, -2.2430, -1.1388]],
             device=torch_device,
         )
-        torch.testing.assert_close(outputs.last_hidden_state[0, :3, :3], expected_slice, rtol=1e-3, atol=1e-3)
+        torch.testing.assert_close(outputs.last_hidden_state[0, :3, :3], expected_slice, rtol=8e-2, atol=8e-2)
```
Comment (qubvel) [inline tests/models/vjepa2/test_modeling_vjepa2.py:270]: fix test on CI (cc @ydshieh)

File: tests/models/vjepa2/test_modeling_vjepa2.py [hunk 5]
```diff
@@ -343,3 +343,22 @@ def test_predictor_partial_mask(self):
         # verify the last hidden states
         expected_shape = torch.Size((1, num_masks, 1024))
         self.assertEqual(outputs.predictor_output.last_hidden_state.shape, expected_shape)
+
+    @slow
+    def test_video_classification(self):
+        checkpoint = "qubvel-hf/vjepa2-vitl-fpc16-256-ssv2"
```
Comment (qubvel) [inline tests/models/vjepa2/test_modeling_vjepa2.py]:
```suggestion
        checkpoint = "facebook/vjepa2-vitl-fpc16-256-ssv2"
```

prepare for checkpoint's transfer

**Comment by qubvel** [comment]
run-slow: vjepa2

File: tests/models/vjepa2/test_modeling_vjepa2.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/vjepa2/test_modeling_vjepa2.py:270]: This looks quite large, especially the dtype is fp32 (?), I guess the original values are not obtained from T4.

Might be better to  change the `expected_slice` instead of change `tol`.

Comment (qubvel) [reply tests/models/vjepa2/test_modeling_vjepa2.py:270]: Then it's going to fail locally 😄 we can choose logits in between, but it still going to be `4e-2`

Comment (qubvel) [reply tests/models/vjepa2/test_modeling_vjepa2.py:270]: It's actually one of the issues. I found it's common to fail integration tests on L4, which are actually pass on T4 on CI. I think it's better to increase tolerance rather than change logits, because we know that the difference is normal for different GPUs

**Comment by qubvel** [comment]
Ok, all slow tests pass now! I will change the checkpoint path to `facebook` now (we will transfer it later)

File: tests/models/vjepa2/test_modeling_vjepa2.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/vjepa2/test_modeling_vjepa2.py:270]: We have now something like `expected = Expectations( {....} )` to have expected values on different hardware, but current we are using it for T4/A10 / AMD ROCM / XPU, **so ok, let's just increase tolerance here**.

(The concern I have here is: sometimes even a diff within 1e-3 could already indicate some real issues)

**Comment by koustuvsinha** [comment]
Looks great to me, thanks for adding this functionality @qubvel !!

File: tests/models/vjepa2/test_modeling_vjepa2.py (hunk 4 above)
Comment (yonigozlan) [reply tests/models/vjepa2/test_modeling_vjepa2.py:270]: I have ~7.2e-05 max diff locally (A10), so this looks to me like a hardware difference issue and not something wrong with the model. 
Although as Yih-Dar suggested it would be great to use `Expectations` with a much lower atol to signal the model is working as expected!

File: src/transformers/models/vjepa2/modeling_vjepa2.py [hunk 6]
```diff
@@ -900,4 +1160,93 @@ def get_vision_features(self, pixel_values_videos) -> torch.Tensor:
         return encoder_output.last_hidden_state
 
 
-__all__ = ["VJEPA2Model", "VJEPA2PreTrainedModel"]
+@auto_docstring(
+    custom_intro="""
+    VJEPA2 Model transformer with a video classification head on top (a linear layer on top of the average pooled hidden
+    states of all tokens) e.g. for ImageNet.
```
Comment (yonigozlan) [inline src/transformers/models/vjepa2/modeling_vjepa2.py]: Could we replace ImageNet with one of the dataset for which we have a model pretrained on?

**Review by yonigozlan — APPROVED** [review]
Great addition thank you @qubvel !
A bit weird for the large atol needed, but it looks like the model might just be quite sensitive to the hardware?

File: tests/models/vjepa2/test_modeling_vjepa2.py (hunk 4 above)
Comment (ydshieh) [reply tests/models/vjepa2/test_modeling_vjepa2.py:270]: Being said that,  I don't know what is the computability capability for L4.

(and we never add expected values for L4 so far 😅 )

**Comment by ydshieh** [comment]
I can update the values for A10/T4 however

File: src/transformers/models/vjepa2/modeling_vjepa2.py (hunk 6 above)
Comment (qubvel) [reply src/transformers/models/vjepa2/modeling_vjepa2.py]: Thanks!

**Comment by qubvel** [comment]
@ydshieh do you have logits values for T4? feel free to push in this/separate PR

**Comment by qubvel** [comment]
Ok, merging this for now. Let's update tests with expectations in a follow-up PR

**Comment by ydshieh** [comment]
Yes I will do, thanks
